stan/math/prim/fun/wiener5_lpdf.hpp

```diff
@@ -0,0 +1,561 @@
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+#ifndef STAN_MATH_PRIM_FUN_WIENER5_LPDF_HPP
+#define STAN_MATH_PRIM_FUN_WIENER5_LPDF_HPP
+
+#include <stan/math/prim/fun/pow.hpp>
+#include <stan/math/prim/fun/square.hpp>
+
+namespace stan {
+namespace math {
+
+/*
+ * The log of the first passage time density function for a (Wiener)
+ * drift diffusion model with 5 parameters.
+ * -boundary separation (alpha),
+ * -drift rate (delta),
+ * -relative starting point (beta),
+ * -non-decision time (tau),
+ * -inter-trial variability in drift rate (sv).
+ * For details see wiener_full_lpdf.
+ */
+
+//-----------------------------------------------
+// DENSITY and derivatives
+//-----------------------------------------------
+
+namespace internal {
+
+// calculate density in log
+template <typename T_y, typename T_alpha, typename T_delta, typename T_beta,
+          typename T_sv>
+return_type_t<T_y, T_alpha, T_delta, T_beta, T_sv> dwiener5(
```

Comment:
Some questions for this one:
1) To make the code more readable we could remove the templates where we do not need them. This is in all sub-functions. They are all called with scalar values (mainly double or integer). We only need templates in the first function in wiener_full_lpdf: wiener_full_prec_impl_lpdf. What do you think?
2) To reduce duplication in wiener5_lpdf we could introduce separate functions that calculate the number of steps for the summation. These are the only parts that should be obsolete to repeat in each function. 
3) Is your suggestion to move all the if-statements that are used to calculate the wiener5 functions to the wiener5_lpdf file into an own function? Then we would also have to move the first 150 lines from wiener_full_prec_impl_lpdf to that new function. So we would have much more code in wiener5_lpdf, but readability could be improved.

Reply:
The suggestion is that you should have the `wiener5` and `wiener7` as their own, standalone, distributions. Then the `wiener_full` distribution is just calling `wiener5_lpdf` or `wiener7_lpdf` with the appropriate arguments (or transformations of)

Reply:
Ok. I can change the code such that the wiener5_lpdf and the wiener7_lpdf files lie in prim/prob and consist of a 'main function' `wiener5_lpdf`, `wiener7_lpdf`, resp., and additionally of all the subfunctions that already exist, and that those functions can be called from wiener_full. (The subfunctions cannot be combined into the 'main function' such that there is only one function in the file.)

I can also remove all the templates that are not needed.

And I can change the ordering of the arguments such that it matches the ordering of `wiener_lpdf`.

So, you want `wiener5` and `wiener7` not to be internal functions anymore? Then the names might be misleading, since with wiener5 you can also compute the 4 parameter diffusion model and with wiener7 you can also compute 6 parameter diffusion models. This is why we made them internal functions such that there is only one function for the user to call (`wiener_full`) and compute multiple diffusion models with this one function.

Reply:
Hey @andrjohns, I changed the structure of the files. I hope that I understood the term 'standalone' correctly ... I'm not sure about this. I still have to move the files into the prim/prob folder and adapt the header and includes. But does the structure now fit to your suggestions?
Maybe the problem occurs that too many lines of code are in one file. Now, in wiener5_lpdf.hpp are 640 and in wiener7_lpdf.hpp are 780 lines of code. How can we solve this? Make another file with all the helper functions as internals?

Reply:
Would you mind if I made the changes I'm thinking and pushed to the branch? I can also make a fork and do it there so you can check first. Just because there are a few Math-specific approaches/tricks that might be a bit easier for me to do and then explain to you, rather than expecting you to be able to do them!

If that works with you I can make those changes later today and put together a brief explainer for you to look at tomorrow?

Reply:
Yes, that would be fine for me. Just push to the branch on my fork.
I just fixed the header issues und moved the two files from fun to prob. I will push them in a minute. Then I can wait for your changes.

Reply:
Great, just ping me here when everything's pushed and then I'll get started

Reply:
Hey @andrjohns, have you any news? I waited with the rearrangement of the tests for your update. Furthermore, I am not sure how to code the mix tests for all my variables.